### PR TITLE
fix: write type definition files (via triple-slash reference) to npm package

### PR DESCRIPTION
**Pull request by dherges**
## I'm submitting a...

```
[x] Bug Fix
[ ] Feature
[ ] Other (Refactoring, Added tests, Documentation, ...)
```

## Checklist

- [ ] Commit Messages follow the [Conventional Commits](https://conventionalcommits.org/) pattern
  - A feature commit message is prefixed "feat:"
  - A bugfix commit message is prefixed "fix:"
- [x] Tests for the changes have been added

## Description

Used to work in 2.0.0-rc.4

Broke in rc5 and later

## Does this PR introduce a breaking change?

```
[ ] Yes
[x] No
```

### CODE REVIEW

**Comment by dherges** [comment]
rc4 and prior:

screen shot 2017-12-25 at 14 37 35
screen shot 2017-12-25 at 14 37 43

**Comment by dherges** [comment]
`ts.createProgram()` on d.ts files in `out` dir and then flatten all `const allNodes = /* collect all nodes ... => */ node.modifers && node.modifiers.find(token => token.kind === ts.SyntaxKind.ExportKeyword)` ... `allNodes.map(node => node.getText()).join(NEWLINE)`

**Comment by dherges** [comment]
What about...? Setting `outDir` to the same directory as the sources and register a custom `WriteFileCallback` on the compilerHost`

https://github.com/Microsoft/TypeScript/blob/6997e9b7316b081f8731d131139c7d7f1bba9092/lib/typescript.d.ts#L2422

**Comment by dherges** [comment]
PoC that it would be possible to pipe/redirect the typescript output through a custom `writeFile` callback

```
$ integration/samples.sh
Building sample ./integration/samples/core/ng-package.json...
Building Angular Package
Building from sources for entry point '@sample/core'
Cleaning build directory
Extracting templateUrl and styleUrls
Processing assets
Inlining templateUrl and styleUrls
Compiling with ngc
Write file output...
... /Users/David/Projects/github/dherges/ng-packagr/integration/samples/core/src/angular.component.js.map
... /Users/David/Projects/github/dherges/ng-packagr/integration/samples/core/.ng_pkg_build/sample-core/out
Write file output...
... /Users/David/Projects/github/dherges/ng-packagr/integration/samples/core/src/angular.component.js
... /Users/David/Projects/github/dherges/ng-packagr/integration/samples/core/.ng_pkg_build/sample-core/out
Write file output...
... /Users/David/Projects/github/dherges/ng-packagr/integration/samples/core/src/angular.component.d.ts
... /Users/David/Projects/github/dherges/ng-packagr/integration/samples/core/.ng_pkg_build/sample-core/out
Write file output...
... /Users/David/Projects/github/dherges/ng-packagr/integration/samples/core/src/angular.directive.js.map
... /Users/David/Projects/github/dherges/ng-packagr/integration/samples/core/.ng_pkg_build/sample-core/out
Write file output...
... /Users/David/Projects/github/dherges/ng-packagr/integration/samples/core/src/angular.directive.js
... /Users/David/Projects/github/dherges/ng-packagr/integration/samples/core/.ng_pkg_build/sample-core/out
Write file output...
... /Users/David/Projects/github/dherges/ng-packagr/integration/samples/core/src/angular.directive.d.ts
... /Users/David/Projects/github/dherges/ng-packagr/integration/samples/core/.ng_pkg_build/sample-core/out
Write file output...
... /Users/David/Projects/github/dherges/ng-packagr/integration/samples/core/src/angular.pipe.js.map
... /Users/David/Projects/github/dherges/ng-packagr/integration/samples/core/.ng_pkg_build/sample-core/out
Write file output...
... /Users/David/Projects/github/dherges/ng-packagr/integration/samples/core/src/angular.pipe.js
... /Users/David/Projects/github/dherges/ng-packagr/integration/samples/core/.ng_pkg_build/sample-core/out
Write file output...
... /Users/David/Projects/github/dherges/ng-packagr/integration/samples/core/src/angular.pipe.d.ts
... /Users/David/Projects/github/dherges/ng-packagr/integration/samples/core/.ng_pkg_build/sample-core/out
Write file output...
... /Users/David/Projects/github/dherges/ng-packagr/integration/samples/core/src/angular.service.js.map
... /Users/David/Projects/github/dherges/ng-packagr/integration/samples/core/.ng_pkg_build/sample-core/out
Write file output...
... /Users/David/Projects/github/dherges/ng-packagr/integration/samples/core/src/angular.service.js
... /Users/David/Projects/github/dherges/ng-packagr/integration/samples/core/.ng_pkg_build/sample-core/out
Write file output...
... /Users/David/Projects/github/dherges/ng-packagr/integration/samples/core/src/angular.service.d.ts
... /Users/David/Projects/github/dherges/ng-packagr/integration/samples/core/.ng_pkg_build/sample-core/out
Write file output...
... /Users/David/Projects/github/dherges/ng-packagr/integration/samples/core/src/angular.module.js.map
... /Users/David/Projects/github/dherges/ng-packagr/integration/samples/core/.ng_pkg_build/sample-core/out
Write file output...
... /Users/David/Projects/github/dherges/ng-packagr/integration/samples/core/src/angular.module.js
... /Users/David/Projects/github/dherges/ng-packagr/integration/samples/core/.ng_pkg_build/sample-core/out
Write file output...
... /Users/David/Projects/github/dherges/ng-packagr/integration/samples/core/src/angular.module.d.ts
... /Users/David/Projects/github/dherges/ng-packagr/integration/samples/core/.ng_pkg_build/sample-core/out
Write file output...
... /Users/David/Projects/github/dherges/ng-packagr/integration/samples/core/public_api.js.map
... /Users/David/Projects/github/dherges/ng-packagr/integration/samples/core/.ng_pkg_build/sample-core/out
Write file output...
... /Users/David/Projects/github/dherges/ng-packagr/integration/samples/core/public_api.js
... /Users/David/Projects/github/dherges/ng-packagr/integration/samples/core/.ng_pkg_build/sample-core/out
Write file output...
... /Users/David/Projects/github/dherges/ng-packagr/integration/samples/core/public_api.d.ts
... /Users/David/Projects/github/dherges/ng-packagr/integration/samples/core/.ng_pkg_build/sample-core/out
Write file output...
... /Users/David/Projects/github/dherges/ng-packagr/integration/samples/core/sample-core.js.map
... /Users/David/Projects/github/dherges/ng-packagr/integration/samples/core/.ng_pkg_build/sample-core/out
Write file output...
... /Users/David/Projects/github/dherges/ng-packagr/integration/samples/core/sample-core.js
... /Users/David/Projects/github/dherges/ng-packagr/integration/samples/core/.ng_pkg_build/sample-core/out
Write file output...
... /Users/David/Projects/github/dherges/ng-packagr/integration/samples/core/sample-core.d.ts
... /Users/David/Projects/github/dherges/ng-packagr/integration/samples/core/.ng_pkg_build/sample-core/out
Writing flat bundle files
Bundling to FESM15
```

```bash
$ git status
Untracked files:
  (use "git add <file>..." to include in what will be committed)

        integration/samples/core/sample-core.metadata.json
```

File: integration/samples/typings/specs/typings.ts [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+    before(() => {
+      VALIDATOR_DTS = fs.readFileSync(
+        path.resolve(__dirname, '..', 'dist', 'validator.d.ts'), 'utf-8');
+    });
+
+    it(`should triple-slash reference to 'src/chalk.d.ts'`, () => {
+      /// <reference path="src/chalk.d.ts" />
+      expect(VALIDATOR_DTS).to.contain('src/chalk.d.ts');
+      expect(VALIDATOR_DTS).to.satisfy(
+        (value: string) => value.startsWith(`/// <reference path="src/chalk.d.ts" />`),
```
Comment (dherges) [inline integration/samples/typings/specs/typings.ts]: Use Case: include additional typings in compilation, write such typings to distributable bundle

Underlying social conflict: Customer lock-in to third-party on version x. Vendor lock-in who won't ship types prior to version x+(something). If possible, one could ship type definitions on an internal npm registry (why are we not doing it?!?)

Proposed solution: setting `compilerOptions.outDir` equal to `compilerOptions.basePath` will output `*.{js,js.map,d.ts}` in the same folder as input `*.ts` files. In a subsequent step, copy `<basePath>/**/*.d.ts` to the distributable bundle - et voila, triple-slash referenced typings are included. In v1, was implemented by copying files on disk (to a temporary working directory).

**Comment by kylecordes** [comment]
If I understand this [BLOCKER] issue right, it should be safe (and potentially helpful) to start testing/using v2 RC in projects where the code doesn't have any triple-slash type references.

**Comment by dherges** [comment]
(rebased to master)

File: src/lib/ng-v5/entry-point/write-package.transform.ts [hunk 2]
```diff
@@ -17,6 +17,7 @@ export const writePackageTransform: Transform = transformFromPromise(async graph
 
   // 5. COPY SOURCE FILES TO DESTINATION
   log.info('Copying staged files');
+  copyFiles(`${path.dirname(ngEntryPoint.entryFilePath)}/**/*.d.ts`, entryPoint.data.outDir);
```
Comment (dherges) [inline src/lib/ng-v5/entry-point/write-package.transform.ts]: Omg, I though the `writeFile` callback was capturing the triple-slash referenced typings. It isn't.

File: src/lib/ts/tsconfig.ts [hunk 3]
```diff
@@ -38,7 +38,7 @@ export const initializeTsConfig = (defaultTsConfig: TsConfig, entryPoint: NgEntr
   tsConfig.options.basePath = basePath;
   tsConfig.options.baseUrl = basePath;
   tsConfig.options.rootDir = basePath;
-  tsConfig.options.outDir = outDir;
+  tsConfig.options.outDir = basePath;
```
Comment (dherges) [inline src/lib/ts/tsconfig.ts]: Seting `outDir`equal to `basePath` just results in a project-relative path for the triple-slash reference in the result d.ts

**Comment by alan-agius4** [comment]
Yay super glad this fix has landed 😁

**Pull request merged**